frankenstein/chunk_store.go

```diff
@@ -0,0 +1,379 @@
+)
+
+var (
+	dynamoRequestDuration = prometheus.NewSummaryVec(prometheus.SummaryOpts{
+		Namespace: "prometheus",
+		Name:      "dynamo_request_duration_seconds",
+		Help:      "Time spent doing DynamoDB requests.",
+	}, []string{"method", "status_code"})
+	dynamoConsumedCapacity = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "prometheus",
+		Name:      "dynamo_consumed_capacity",
+		Help:      "The capacity units consumed by operation.",
+	}, []string{"method"})
+	s3RequestDuration = prometheus.NewSummaryVec(prometheus.SummaryOpts{
+		Namespace: "prometheus",
+		Name:      "s3_request_duration_seconds",
+		Help:      "Time spent doing S3 requests.",
+	}, []string{"method", "status_code"})
+)
+
+func init() {
+	prometheus.MustRegister(dynamoRequestDuration)
+	prometheus.MustRegister(dynamoConsumedCapacity)
+	prometheus.MustRegister(s3RequestDuration)
+}
+
+type Chunk struct {
+	ID      string       `json:"-"`
+	From    model.Time   `json:"from"`
+	Through model.Time   `json:"through"`
+	Metric  model.Metric `json:"metric"`
+	Data    []byte       `json:"-"`
+}
+
+type ChunkStore interface {
+	Put([]Chunk) error
+	Get(from, through model.Time, matchers ...*metric.LabelMatcher) ([]Chunk, error)
+}
+
+type ChunkStoreConfig struct {
+	S3URL       string
+	DynamoDBURL string
+}
+
+func NewChunkStore(cfg ChunkStoreConfig) (ChunkStore, error) {
+	s3URL, err := url.Parse(cfg.S3URL)
+	if err != nil {
+		return nil, err
+	}
+
+	s3Config, err := awsConfigFromURL(s3URL)
+	if err != nil {
+		return nil, err
+	}
+
+	dynamodbURL, err := url.Parse(cfg.DynamoDBURL)
+	if err != nil {
+		return nil, err
+	}
+
+	dynamoDBConfig, err := awsConfigFromURL(dynamodbURL)
+	if err != nil {
+		return nil, err
+	}
+
+	tableName := strings.TrimPrefix(dynamodbURL.Path, "/")
+	bucketName := strings.TrimPrefix(s3URL.Path, "/")
+
+	return &chunkStore{
+		dynamodb:   dynamodb.New(session.New(dynamoDBConfig)),
+		s3:         s3.New(session.New(s3Config)),
+		tableName:  tableName,
+		bucketName: bucketName,
+	}, nil
+}
+
+func awsConfigFromURL(url *url.URL) (*aws.Config, error) {
+	if url.User == nil {
+		return nil, fmt.Errorf("Must specify username & password in URL")
+	}
+	password, _ := url.User.Password()
+	creds := credentials.NewStaticCredentials(url.User.Username(), password, "")
+	config := aws.NewConfig().WithCredentials(creds)
+	if strings.Contains(url.Host, ".") {
+		config = config.WithEndpoint(fmt.Sprintf("http://%s", url.Host)).WithRegion("dummy")
+	} else {
+		config = config.WithRegion(url.Host)
+	}
+	return config, nil
+}
+
+type chunkStore struct {
+	dynamodb   *dynamodb.DynamoDB
+	s3         *s3.S3
+	tableName  string
+	bucketName string
+	cfg        ChunkStoreConfig
+}
+
+func hours(from, through model.Time) []model.Time {
+	// TODO it
+	return nil
+}
+
+func (c *chunkStore) Put(chunks []Chunk) error {
+	// TODO: parallelise
+	for _, chunk := range chunks {
+		err := timeRequest("Put", s3RequestDuration, func() error {
```

Comment:
Ah, with the other calls I see that this doesn't refer to the HTTP method, but the S3 operation. Should the `method` label on the histogram be renamed to `operation` then? Or whatever the right term is.